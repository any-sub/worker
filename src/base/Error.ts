export abstract class WorkerError extends Error {
  public static INTERNAL = "INTERNAL";

  public abstract get code(): string;
}

export class ElementNotFoundError extends WorkerError {
  public get code(): string {
    return "ELEMENT_NOT_FOUND";
  }

  public constructor() {
    super("Element not found");
  }
}

export class UnhandledSourceTypeError extends WorkerError {
  public get code(): string {
    return "UNHANDLED_SOURCE_TYPE";
  }

  public constructor() {
    super("Unhandled source type");
  }
}

export class MaxRetryReachedError extends WorkerError {
  public get code(): string {
    return "MAX_RETRY_REACHED";
  }

  public constructor() {
    super("Max retry reached");
  }
}

export class ContentTypeMismatchError extends WorkerError {
  public get code(): string {
    return "CONTENT_TYPE_MISMATCH";
  }

  public constructor() {
    super("Content type returned does not match");
  }
}

export class ResultNotOkError extends WorkerError {
  public get code(): string {
    return "RESULT_NOT_OK";
  }

  public constructor(contents: string) {
    super("Result not ok: " + contents);
  }
}
