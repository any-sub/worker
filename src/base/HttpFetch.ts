/* istanbul ignore file */
import { Injectable } from "@tsed/di";
import nodeFetch, { RequestInfo, RequestInit, Response } from "node-fetch";

@Injectable()
export class HttpFetch {
  public async fetch(url: RequestInfo, init?: RequestInit): Promise<Response> {
    return nodeFetch(url, init);
  }
}
