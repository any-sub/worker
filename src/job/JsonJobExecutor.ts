import { JobExecutor } from "./JobExecutor";
import { Work } from "@any-sub/worker-transport";
import { ResultReport } from "../model/Report";
import { JsonReader } from "../readers";
import { JsonConsumer } from "../consumers/JsonConsumer";

export class JsonJobExecutor extends JobExecutor {
  constructor(private readonly reader: JsonReader, private readonly consumer: JsonConsumer) {
    super();
  }

  protected async executeInternal(work: Work): Promise<ResultReport[]> {
    const source = await this.reader.read(work.source.location);
    return this.consumer.consume(source, work);
  }
}
