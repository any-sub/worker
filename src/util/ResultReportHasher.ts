import { Injectable } from "@tsed/di";
import { ResultReport } from "../model/Report";
import { createHash } from "crypto";

@Injectable()
export class ResultReportHasher {
  public hash(unit: ResultReport): string {
    return createHash("md5").update(JSON.stringify(unit)).digest("hex");
  }
}
