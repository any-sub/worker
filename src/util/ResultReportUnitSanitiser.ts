import { Injectable } from "@tsed/di";
import { ResultReport } from "../model/Report";

@Injectable()
export class ResultReportUnitSanitiser {
  public sanitise(unit: ResultReport, baseURL: string): ResultReport {
    return {
      ...this.sanitiseField("title", 250, unit.title),
      ...this.sanitiseField("description", 500, unit.description),
      ...this.sanitiseURL(baseURL, "image", unit.image),
      ...this.sanitiseURL(baseURL, "url", unit.url)
    };
  }

  private sanitiseField(fieldName: keyof ResultReport, maxLength: number, fieldValue?: Nullable<string>): ResultReport | undefined {
    return fieldValue
      ? {
          [fieldName]: this.sanitiseValue(fieldValue).substring(0, maxLength)
        }
      : undefined;
  }

  private makeURL(baseURL: string, pathRaw: string): string {
    const path = this.sanitiseValue(pathRaw);
    let url;
    try {
      url = new URL(path);
    } catch (_) {
      url = new URL(path, baseURL);
    }
    return url.href;
  }

  private sanitiseURL(baseURL: string, fieldName: keyof ResultReport, fieldValue?: Nullable<string>): ResultReport | undefined {
    return fieldValue
      ? {
          [fieldName]: this.makeURL(baseURL, this.sanitiseValue(fieldValue)).substring(0, 2048)
        }
      : undefined;
  }

  private sanitiseValue(value: string) {
    return value.trim();
  }
}
