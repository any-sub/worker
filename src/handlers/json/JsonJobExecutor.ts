import { JobExecutor } from "../../job/JobExecutor";
import { Work } from "@any-sub/worker-transport";
import { ResultReport } from "../../model/Report";
import { JsonConsumer } from "./JsonConsumer";
import { Injectable } from "@tsed/di";
import { JsonReader } from "./JsonReader";

@Injectable()
export class JsonJobExecutor extends JobExecutor {
  constructor(
    private readonly reader: JsonReader,
    private readonly consumer: JsonConsumer
  ) {
    super();
  }

  protected async executeInternal(work: Work): Promise<ResultReport[]> {
    const source = await this.reader.read(work.source.location);
    return this.consumer.consume(source, work);
  }
}
