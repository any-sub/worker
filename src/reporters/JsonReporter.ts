import { Injectable } from "@tsed/di";
import { AbstractReportUnit, Reporter } from "./Reporter";
import { Report, TextReporting } from "@any-sub/worker-transport";
import { ResultReport } from "../model/Report";
import { JsonSource } from "../readers";
import { isPresent } from "../util/TypeUtils";

@Injectable()
export class JsonReporter extends Reporter<JsonSource> {
  public buildReport(content: ReportUnit[], report?: Report): ResultReport[] {
    return report ? this.build(content, report) : this.buildDefault(content);
  }

  private build(content: ReportUnit[], report: Report): ResultReport[] {
    return content.map((unit) => ({
      title: this.buildText(unit.element, report?.title, unit.title),
      description: this.buildText(unit.element, report?.description, unit.description),
      image: this.buildText(unit.element, report?.image, unit.image),
      url: this.buildText(unit.element, report?.url, unit.url)
    }));
  }

  private buildDefault(content: ReportUnit[]): ResultReport[] {
    return content
      .map((el) => this.stringify(el.description ?? el.element))
      .filter(isPresent)
      .map((description) => ({ description }));
  }

  private stringify(source: JsonSource): string {
    return typeof source === "string" ? source : JSON.stringify(source);
  }

  private buildText(element: JsonSource, report?: TextReporting, reportingElement?: JsonSource) {
    return this.reportTextElement(reportingElement ?? element, report);
  }

  private reportTextElement(element: JsonSource, options?: TextReporting) {
    const stringValue = this.stringify(element);
    const groups = (options?.match && stringValue?.match(options.match)?.groups) || {};
    return this.reportText(options?.template ?? stringValue, { ...{ $: element }, ...groups });
  }
}

export type ReportUnit = AbstractReportUnit<JsonSource>;
