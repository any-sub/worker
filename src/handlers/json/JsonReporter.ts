import { Injectable } from "@tsed/di";
import { AbstractReportUnit, PartReporter, Reporter } from "../../reporters/Reporter";
import { Report, TextReporting } from "@any-sub/worker-transport";
import { JsonSource } from "../../readers";

@Injectable()
export class JsonReporter extends Reporter<JsonSource> {
  getDefaultReporter(): PartReporter<JsonSource> {
    return new TextReporter("title");
  }

  getTitleReporter(): PartReporter<JsonSource> {
    return new TextReporter("title");
  }

  getDescriptionReporter(): PartReporter<JsonSource> {
    return new TextReporter("description");
  }

  getImageReporter(): PartReporter<JsonSource> {
    return new TextReporter("image");
  }

  getUrlReporter(): PartReporter<JsonSource> {
    return new TextReporter("url");
  }
}

export type ReportUnit = AbstractReportUnit<JsonSource>;

class TextReporter extends PartReporter<JsonSource> {
  private readonly part: keyof Report;

  constructor(part: keyof Report) {
    super();
    this.part = part;
  }

  public build(element?: JsonSource, report?: Report, reportingElement?: JsonSource) {
    if (!element) return null;
    return this.buildText(element ?? reportingElement, report?.[this.part]);
  }

  private stringify(source: JsonSource): string {
    return typeof source === "string" ? source : JSON.stringify(source);
  }

  private buildText(element: JsonSource, report?: Nullable<TextReporting>, reportingElement?: JsonSource) {
    return this.reportTextElement(reportingElement ?? element, report);
  }

  private reportTextElement(element: JsonSource, options?: Nullable<TextReporting>) {
    const stringValue = this.stringify(element);
    const groups = (options?.match && stringValue?.match(options.match)?.groups) || {};
    return element ? this.reportText(options?.template ?? "", { ...{ $: element }, ...groups }) : "";
  }
}
