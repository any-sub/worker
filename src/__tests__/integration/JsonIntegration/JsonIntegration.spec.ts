import { JsonReader } from "../../../readers";
import { expect } from "@jest/globals";
import { MockServer } from "jest-mock-server";
import { HttpFetch } from "../../../base";
import * as fs from "fs";
import * as path from "path";
import { JsonConsumer } from "../../../consumers";
import { JsonReporter } from "../../../reporters/JsonReporter";
import { JsonJobExecutor } from "../../../job/JsonJobExecutor";
import { ResultReportUnitSanitiser } from "../../../util/ResultReportUnitSanitiser";
import { ResultReportHasher } from "../../../util/ResultReportHasher";

const readJsonSource = (name: string): string => {
  return fs.readFileSync(path.join(__dirname, `${name}.json`), "utf-8");
};

describe("JsonJobExecutor integration", () => {
  let executor: JsonJobExecutor;
  const server = new MockServer();
  beforeAll(() => server.start());
  afterAll(() => server.stop());
  beforeEach(() => {
    server.reset();
    const reader = new JsonReader(new HttpFetch());
    const consumer = new JsonConsumer(new JsonReporter());
    executor = new JsonJobExecutor(reader, consumer);
    executor.sanitiser = new ResultReportUnitSanitiser();
    executor.hasher = new ResultReportHasher();
  });

  it("should throw for single string", async () => {
    // Given
    server.get("/").mockImplementation((ctx) => {
      ctx.status = 200;
      ctx.response.set({ "content-type": "application/json" });
      ctx.body = readJsonSource("single-string");
    });

    // When - Then
    await expect(
      executor.execute({
        source: {
          location: `${server.getURL().toString()}`,
          type: "json"
        },
        type: "http",
        id: "",
        consume: {}
      })
    ).rejects.toThrow("obj needs to be an object");
  });

  it("should report with default reporting for object", async () => {
    // Given

    server.get("/").mockImplementation((ctx) => {
      ctx.status = 200;
      ctx.response.set({ "content-type": "application/json" });
      ctx.body = readJsonSource("simple-object");
    });

    // When
    const result = (await executor.execute({
      source: {
        location: `${server.getURL().toString()}`,
        type: "json"
      },
      type: "http",
      id: "",
      consume: {}
    })) as any;

    // Then
    expect(result.data[0].description.trim()).toEqual(`{"values":[{"foo":"bar"}]}`);
  });

  it("should report with complex reporting", async () => {
    // Given
    server.get("/").mockImplementation((ctx) => {
      ctx.status = 200;
      ctx.response.set({ "content-type": "application/json" });
      ctx.body = readJsonSource("complex-object");
    });

    // When
    const result = await executor.execute({
      source: {
        location: `${server.getURL().toString()}`,
        type: "json"
      },
      type: "http",
      id: "",
      consume: {
        lookup: {
          mode: "jsonpath",
          value: "$.details.updates.*"
        },
        parts: {
          title: {
            mode: "jsonpath",
            value: "$.heading"
          },
          description: {
            mode: "jsonpath",
            value: "$.body"
          },
          url: {
            mode: "jsonpath",
            value: "$.id"
          },
          image: {
            mode: "jsonpath",
            value: "$.id"
          }
        }
      },
      report: {
        title: {
          template: "{{$}}"
        },
        description: {
          template: `{{second}} {{first}}`,
          match: "^(?<first>\\w+) with some (?<second>\\w+) value$"
        },
        image: {
          template: "http://someurl/{{$}}.png"
        },
        url: {
          template: "http://someurl/{{$}}"
        }
      }
    });

    // Then
    expect(result!.data).toMatchSnapshot();
  });
});
