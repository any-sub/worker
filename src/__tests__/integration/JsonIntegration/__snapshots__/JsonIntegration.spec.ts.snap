// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JsonJobExecutor integration should report with complex reporting 1`] = `
[
  {
    "description": "foo Desc1",
    "hash": "1d4c68db06077149844f3b81dbd3cd37",
    "image": "http://someurl/14533.png",
    "title": "Title1",
    "url": "http://someurl/14533",
  },
  {
    "description": "bar Desc2",
    "hash": "ce5b1cf2b056d5537603c3e65bf839a7",
    "image": "http://someurl/22554.png",
    "title": "Title2",
    "url": "http://someurl/22554",
  },
]
`;
